fish-rust/src/color.rs

```diff
@@ -0,0 +1,453 @@
+use std::{array, cmp::Ordering, mem::size_of};
+
+use modular_bitfield::{bitfield, specifiers::B24, BitfieldSpecifier};
+
+use crate::{
+    wchar::{widestrs, wstr, WExt, WString, L},
+    wutil::sprintf,
+};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Color24 {
+    r: u8,
+    g: u8,
+    b: u8,
+}
+
+impl Color24 {
+    fn from_bits(bits: u32) -> Self {
+        assert_eq!(bits >> 24, 0, "from_bits() called with non-zero high byte");
+
+        Self {
+            r: (bits >> 16) as u8,
+            g: (bits >> 8) as u8,
+            b: bits as u8,
+        }
+    }
+}
+
+#[derive(BitfieldSpecifier, Debug, Copy, Clone, PartialEq, Eq)]
+#[bits = 3]
+enum Type {
+    None,
+    Named,
+    Rgb,
+    Normal,
+    Reset,
+}
+
+#[allow(unused_braces)] // triggered inside modular_bitfield generated code
+mod dummy {
+    use modular_bitfield::bitfield;
+
+    #[bitfield(filled = false)]
+    #[derive(BitfieldSpecifier, Debug, Copy, Clone, PartialEq, Eq)]
+    pub struct Flags {
+        pub bold: bool,
+        pub underline: bool,
+        pub italics: bool,
+        pub dim: bool,
+        pub reverse: bool,
+    }
+}
+use dummy::Flags;
+
+#[bitfield]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct RgbColor {
+    #[bits = 3]
+    typ: Type,
+    #[bits = 5]
+    flags: Flags,
+    /// Either the R, G and B bytes for [`Type::Rgb`], or the name index for [`Type::Named`].
+    data: B24,
+}
+
+const _: () = assert!(size_of::<RgbColor>() <= 4, "RgbColor is too big");
```

Comment:
Thank you for faithfully propagating this assert, but let's remove it, this isn't important.